src/main/java/org/candlepin/model/ConsumerCloudData.java

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright (c) 2009 - 2024 Red Hat, Inc.
+ *
+ * This software is licensed to you under the GNU General Public License,
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+
+package org.candlepin.model;
+
+import org.candlepin.util.Util;
+
+import org.hibernate.annotations.GenericGenerator;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.FetchType;
+import javax.persistence.GeneratedValue;
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.OneToOne;
+import javax.persistence.Table;
+import javax.validation.constraints.NotNull;
+
+@Entity
+@Table(name = ConsumerCloudData.DB_TABLE)
+public class ConsumerCloudData extends AbstractHibernateObject<ConsumerCloudData> {
+
+    /** Name of the table backing this object in the database */
+    public static final String DB_TABLE = "cp_consumer_cloud_data";
+
+    /** Max length for a value in the ID field */
+    public static final int ID_MAX_LENGTH = 32;
+    /** Max length for a value in the UUID field */
+    public static final int UUID_MAX_LENGTH = 36;
+    /** Max length for a value in the cloud account ID field */
+    public static final int CLOUD_ACCOUNT_ID_MAX_LENGTH = 255;
+    /** Max length for a value in the cloud instance ID field */
+    public static final int CLOUD_INSTANCE_ID_MAX_LENGTH = 170;
+    /** Max length for a value in the cloud offering ID field */
+    public static final int CLOUD_OFFERING_ID_MAX_LENGTH = 170;
```

Comment:
CLOUD_OFFERING_ID_MAX_LENGTH needs to be updated to 512

Reply:
the 512 is the maximum size of the varchar in the table because the offers can be 2 so it will have 170 + 170 +1 so we would not fit into 255

Reply:
That is my bad, and I see what you are saying! I thought this dictating the max length of the cloud offering ID field and not the individual offers. Could we update the comment on this static variable to reflect that this is the max length of a cloud offering and not the whole cloud offering ID field?

Reply:
What exactly is an "offering ID" and how does it relate to product IDs? Do we have the specifications of what these fields can possibly contain defined anywhere? It's possible these are oversized by almost an order of magnitude.

Reply:
Sorry, I just now saw this comment. The cloud offering ID is the cloud marketplace product identifier that is included in the cloud instance metadata and correlates to the cloud offering that was bought on the cloud marketplace. Cloudoscope maps these cloud offering Ids to Red Hat Product IDs.

For AWS, the metadata has a billing_product and a marketplace_product field that contains these cloud product identifiers. I included some links below to what these strings look like, but here are some examples (not real ones):

`"billing_product"="bp-6da53045,bp-6554f011,bp-3bd53007"`
`"marketplace_product"="961yclsj3tip2h37aja4sdtkla"`

Here is a link to a discussion on this from the design doc:
https://docs.google.com/document/d/1xt-XHGBgK6t3j6hfZEv2966bGWUeVLfOWaLn2Rl6ne4/edit?disco=AAABQH4Gep0

Here are some examples from cloud image metadata investigation by Nikos:
https://docs.google.com/spreadsheets/d/1HzC3S2eZpTrljgklk4ryj-hmt_71jCtj2pfV7RC0TrY/edit?gid=0#gid=0

Reply:
Sorry, I just now saw this comment. The cloud offering ID is the cloud marketplace product identifier that is included in the cloud instance metadata and correlates to the cloud offering that was bought on the cloud marketplace. Cloudoscope maps these cloud offering Ids to Red Hat Product IDs.

For AWS, the metadata has a billing_product and a marketplace_product field that contains these cloud product identifiers. I included some links below to what these strings look like, but here are some examples (not real ones):

`"billing_product"="bp-6da53045,bp-6554f011,bp-3bd53007"`
`"marketplace_product"="961yclsj3tip2h37aja4sdtkla"`

Here is a link to a discussion on this from the design doc:
https://docs.google.com/document/d/1xt-XHGBgK6t3j6hfZEv2966bGWUeVLfOWaLn2Rl6ne4/edit?disco=AAABQH4Gep0

Here are some examples from cloud image metadata investigation by Nikos:
https://docs.google.com/spreadsheets/d/1HzC3S2eZpTrljgklk4ryj-hmt_71jCtj2pfV7RC0TrY/edit?gid=0#gid=0

Reply:
The 170 character length Stepan mentioned is coming from the `cp_anonymous_cloud_consumers.cloud_offering_id` column as that is what we set the length to be. Unfortunately I can't find any documentation on if we were anticipating on storing a comma separated list of Cloud Offering IDs when we decided on that length.

Either way the 512 length for the column is too large. We could set the length to be 170 to match `cp_anonymous_cloud_consumers.cloud_offering_id`. Based on the longest examples we have, and concatenating the `billing_product` and `marketplace_product` field together, this would be more than enough.

Link to the PR where `cp_anonymous_cloud_consumers.cloud_offering_id` was created:
https://github.com/candlepin/candlepin/pull/4154

Reply:
I think we should be good reducing the limit to match `cp_anonymous_cloud_consumers.cloud_offering_id` at 170 length.

Based on the information we currently have on our 1P and Gold images, the longest string is only 62 characters long if you concatenate the `billing_product` and `marketplace_product` fields. So, we would have to triple that size for us to run into issues.

Reply:
Update: After a discussion today, we decided to move the size of this column to 255